crates/sui-graphql-rpc/src/server/data_provider.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// For testing, use existing RPC as data source
+
+use std::str::FromStr;
+
+use crate::types::address::Address;
+use crate::types::balance::Balance;
+use crate::types::base64::Base64;
+use crate::types::big_int::BigInt;
+use crate::types::transaction_block::TransactionBlock;
+use crate::types::{object::Object, sui_address::SuiAddress};
+use async_graphql::*;
+use sui_json_rpc_types::{
+    SuiObjectDataOptions, SuiPastObjectResponse, SuiTransactionBlockResponseOptions,
+};
+use sui_sdk::types::base_types::ObjectID;
+use sui_sdk::types::digests::TransactionDigest;
+use sui_sdk::SuiClient;
+
+pub(crate) async fn fetch_obj(
+    cl: &SuiClient,
+    address: SuiAddress,
+    version: Option<u64>,
+) -> Result<Option<Object>> {
+    let oid: ObjectID = address.to_array().as_slice().try_into()?;
+    let opts = SuiObjectDataOptions::full_content();
+
+    let g = match version {
+        Some(v) => match cl
+            .read_api()
+            .try_get_parsed_past_object(oid, v.into(), opts)
+            .await?
+        {
+            SuiPastObjectResponse::VersionFound(x) => x,
+            _ => return Ok(None),
+        },
+        None => {
+            let val = cl.read_api().get_object_with_options(oid, opts).await?;
+            if val.error.is_some() || val.data.is_none() {
+                return Ok(None);
+            }
+            val.data.unwrap()
+        }
+    };
+    Ok(Some(convert_obj(g)))
+}
+
+pub(crate) async fn fetch_balance(
+    cl: &SuiClient,
+    address: &SuiAddress,
```

Comment:
SuiAddress implements copy and clone - could we take the address instead of a reference and pass that directly to coin_read_api().get_balance(), since get_balance() takes a SuiAddress?

Reply:
less data to move around